feat: add bookings and comments functionality

Alex-M-Markov opened the pull request.

Alex-M-Markov committed:
feat: add bookings and comments functionality

kukolev commented on src/main/java/shareit/app/booking/BookingRepository.java:
```diff
@@ -0,0 +1,52 @@
+    Booking findFirstByItem_IdAndStartIsAfterAndStatusIs(Long itemId, LocalDateTime current,
+        BookingStatus status, Long userId);
+}
+/*
```
Закомментированный код нужно удалить или раскомментировать и отладить. Оставлять закомментированный код это плохая практика, т.к. это только запутает программиста, который будет анализировать код.

kukolev commented on src/main/java/shareit/app/item/CommentDto.java:
```diff
@@ -0,0 +1,23 @@
+    private String text;
+    private Item item;
+    private User author;
+    LocalDateTime created;
```
Нарушен принцип инкапсуляции. Поля класса не должны быть видимы другим классам. Лучше сделать поле приватным и сделать геттер и сеттер для полей, если нужен доступ извне.

kukolev commented on src/main/java/shareit/app/item/CommentDtoToReturn.java:
```diff
@@ -0,0 +1,17 @@
+    private Long id;
+    private String text;
+    private String authorName;
+    LocalDateTime created;
```
Нарушен принцип инкапсуляции. Поля класса не должны быть видимы другим классам. Лучше сделать поле приватным и сделать геттер и сеттер для полей, если нужен доступ извне.

kukolev commented on src/main/java/shareit/app/item/CommentMapper.java:
```diff
@@ -0,0 +1,18 @@
+
+import lombok.Data;
+
+@Data
```
По моему эта аннотация лишняя. Она не нужна в маппере.

kukolev commented on src/main/java/shareit/app/item/ItemRepository.java:
```diff
@@ -0,0 +1,19 @@
+
+    List<Item> findAllByOwnerEquals(User user);
+
+    @Query(" select i from Item i " +
```
Лишний пробел перед select и ниже перед or. Это мелочи и ни на что не влияет, но по моему лучше поправить. Так код будет выглядеть аккуратней.

kukolev commented on src/main/java/shareit/app/item/ItemServiceImpl.java:
```diff
@@ -54,35 +70,105 @@ private ItemDto updateItemFields(Long id, ItemDto item) {
+        try {
+            return ItemMapper.toItemDto(itemRepository.getReferenceById((itemId)));
+        } catch (EntityNotFoundException e) {
+            throw new ItemNotFoundException();
```
При выбросе эксепшенов лучше передавать пойманный эксепшен как параметр - `throw new ItemNotFoundException(e)`. Тогда инфомация о стектрейсе не потеряется.

Также нужно поправить везде в коде.

Alex-M-Markov committed:
fix: add privacy to Comments entities' variables
style: delete excessive blanks in SQL methods
feat: add Throwable causes to Exceptions for better logging

kukolev commented on src/main/java/shareit/app/item/CommentRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface CommentRepository extends JpaRepository<Comment, Long> {
+
+    List<Comment> findAllByItem_Id(Long itemId, Sort sort);
```
Нарушение кодстайла: название методов должно быть выполнено в стиле camelCase.

kukolev commented on src/main/java/shareit/app/item/ItemServiceImpl.java:
```diff
@@ -54,35 +70,106 @@ private ItemDto updateItemFields(Long id, ItemDto item) {
+    @Override
+    public ItemDtoWithBookings getItemByIdWithBookings(Long userId, Long itemId) {
+        try {
+            System.out.println(
```
Логирование нужно делать через логгер, а не просто выводить текст в консоль, т.к. тогда можно будет управлять настройками логирования. Думаю этот код остался после отладки. Нужно поправить.

kukolev commented on src/main/java/shareit/app/booking/BookingServiceImpl.java:
```diff
@@ -0,0 +1,156 @@
+
+@Service
+@RequiredArgsConstructor
+public class BookingServiceImpl implements BookingService {
```
Логику сервисов лучше покрыть логами. Так будет проще анализировать поведение программы, если что-то пойдет не так.

kukolev commented on src/main/java/shareit/app/item/ItemServiceImpl.java:
```diff
@@ -1,40 +1,56 @@
 @Service
-@AllArgsConstructor
+@RequiredArgsConstructor
 public class ItemServiceImpl implements ItemService {
```
Логику сервисов лучше покрыть логами. Так будет проще анализировать поведение программы, если что-то пойдет не так.

kukolev commented on src/main/java/shareit/app/user/UserServiceImpl.java:
```diff
@@ -3,26 +3,27 @@
 @Service
-@AllArgsConstructor
+@RequiredArgsConstructor
 public class UserServiceImpl implements UserService {
```
Логику сервисов лучше покрыть логами. Так будет проще анализировать поведение программы, если что-то пойдет не так.

Alex-M-Markov committed:
feat: add logging to all Service classes
feat: add UnsupportedStatusException to fit to POSTMAN tests
fix: Get current bookings query

kukolev commented on src/main/java/shareit/app/booking/BookingController.java:
```diff
@@ -0,0 +1,76 @@
+        @RequestHeader(USER_HEADER) Long userId,
+        @RequestParam(name = "state", required = false, defaultValue = "ALL")
+        BookingIncomingStates bookingState) throws UnsupportedStatusException {
+        if (bookingState.equals(BookingIncomingStates.UNSUPPORTED_STATUS)) {
```
Это часть валидации входных данных. Валидация это часть бизнес-логики и должна быть перенесена в сервис.

kukolev commented on src/main/java/shareit/app/booking/BookingController.java:
```diff
@@ -0,0 +1,76 @@
+        @RequestHeader(USER_HEADER) Long userId,
+        @RequestParam(name = "state", required = false, defaultValue = "ALL")
+        BookingIncomingStates bookingState) throws UnsupportedStatusException {
+        if (bookingState.equals(BookingIncomingStates.UNSUPPORTED_STATUS)) {
```
Это часть валидации входных данных. Валидация это часть бизнес-логики и должна быть перенесена в сервис.

kukolev commented on src/main/java/shareit/app/booking/BookingRepository.java:
```diff
@@ -0,0 +1,52 @@
+
+public interface BookingRepository extends JpaRepository<Booking, Long> {
+
+    List<Booking> findByBooker_Id(Long bookerId, Sort sort);
```
Нарушение кодстайла: название методов должно быть выполнено в стиле camelCase.

Alex-M-Markov committed:
feat: Move booking checking methods from Controller to Service
style: Adapt BookingRepository methods to camelCase
